File: CigaretteSmokers/CigaretteSmokers/ingredients.cpp
#include "ingredients.h"

using namespace std;

ingredients::ingredients(bool tabacco /*= false*/, bool paper /*= false*/, bool matches /*= false*/)
    : m_tabacco(tabacco)
    , m_paper(paper)
    , m_matches(matches)
{

}


void ingredients::wait_for_complete_with_internal_supply(ingredients::infinit_t internal_supply)
{
    unique_lock<mutex> guard(m_mt);
    m_cv_supply_ready.wait(guard, [&]() {
        switch (internal_supply) {
        case ingredients::infinit_t::tabacco:
            return m_paper && m_matches;
        case ingredients::infinit_t::paper:
            return m_tabacco && m_matches;
        case ingredients::infinit_t::matches:
            return m_paper && m_tabacco;
        }
    });
}



void ingredients::reset()
{
    unique_lock<mutex> guard(m_mt);
    m_tabacco = false;
    m_paper = false;
    m_matches = false;
    m_cv_supply_used.notify_one();
}

File: CigaretteSmokers/CigaretteSmokers-Go/ingredients.go
package main

import (
	"fmt"
	"sync"
)

var mutex sync.Mutex
var cond_supply_ready = sync.NewCond(&mutex)
var cond_supply_used = sync.NewCond(&mutex)

type ingredients_t int

const (
	TABACCO ingredients_t = iota
	PAPER
	MATCHES
)

type ingredients struct {
	tabacco bool
	paper   bool
	matches bool
}

func makeIngredients() ingredients {
	return ingredients{false, false, false}
}

func (g ingredients) complete_with_internal_supply(internal_supply ingredients_t) bool {
	fmt.Println("T:", g.tabacco, ", P:", g.paper, ", M:", g.matches, ", infinit:", internal_supply)
	switch internal_supply {
	case TABACCO:
		return g.matches && g.paper
	case PAPER:
		return g.tabacco && g.matches
	case MATCHES:
		return g.tabacco && g.paper
	}
	return false
}

func (g ingredients) wait_for_complete_with_internal_supply(internal_supply ingredients_t) {
	cond_supply_ready.L.Lock()
	defer cond_supply_ready.L.Unlock()
	for !g.complete_with_internal_supply(internal_supply) {
		cond_supply_ready.Wait()
	}
}

func (g ingredients) reset() {
	mutex.Lock()
	defer mutex.Unlock()
	g.tabacco = false
	g.paper = false
	g.matches = false
	cond_supply_used.Broadcast()
}
